File: Cpp/703-Kth-Largest-Element-in-a-Stream.cpp
// https://leetcode.com/problems/kth-largest-element-in-a-stream/

#include <bits/stdc++.h>
using namespace std;

class KthLargest {
    /*
    Methodology | 
    From init, create max heap from given stream, pop kth largest items
        into a separate min heap. 
        
    If there are less than k items at init, 
        add will require a kth item by constraint.
        Therefore we can add a 'dummy' value to the top of the min heap for the first add
    
    At every add, compare value to top of heap, 
        if less or eq, return top,
        if greater, insert value, pop first top and return new top.
    */
    
private:
    vector<int> heap; // heap of at most k items
    
public:
    KthLargest(int k, vector<int>& nums) {
        make_heap(nums.begin(),nums.end());
        int n = nums.size();
        for (int i = 0; i < k and i < n; i++) {
            heap.push_back(nums.front());
            
            pop_heap(nums.begin(),nums.end());
            nums.pop_back();  
        }
        
        // operator> as greater<int>() for min-heap
        make_heap(heap.begin(),heap.end(),greater<int>());
        
        // If there are < k items in heap, add kth item as '-inf' (-10001)
        if (heap.size() < k) {
            heap.push_back(-10001);
            push_heap(heap.begin(),heap.end(),greater<int>());
        }
    }
    
    int add(int val) {
        int top = heap.front();
        if (val <= top) {
            return  top;
        }
        
        // push val
        heap.push_back(val);
        push_heap(heap.begin(),heap.end(), greater<int>());
        
        // pop top return new top
        pop_heap(heap.begin(),heap.end(), greater<int>());
        heap.pop_back();
        
        return heap.front();
    }
};

File: Python/703-Kth-Largest-Element-in-a-Stream.py
# https://leetcode.com/problems/kth-largest-element-in-a-stream/

import heapq
from typing import * 

class KthLargest:
    """
    Methodology:
    
    On init, create a minheap from the stream.
        If the size of the stream is less than k, add '-inf' (-10001) to act as a dummy head
        Otherwise, pop elements until the heap is k large. 
        The top of the heap will be the kth largest value
        
    On add, compare val to the top of the heap.
        If val is less or eq to top, do nothing return top.
        Otherwise push new val, pop old head, return new head.
    """
    def __init__(self, k: int, nums: List[int]):
        self.q = nums
        heapq.heapify(self.q)
        
        # shrink heap to k big
        while len(self.q) > k:
            heapq.heappop(self.q)
            
        # if stream is smaller than k, add dummy value.
        # next add is guarenteed to increase the heap to k large
        if len(self.q) < k:
            heapq.heappush(self.q,-10001)


    def add(self, val: int) -> int:
        # if the value is less than current kth largest, do nothing
        if val <= self.q[0]:
            return self.q[0]
        
        # push new value, pop old one to maintain k size heap
        heapq.heappush(self.q,val)
        heapq.heappop(self.q)
        
        return self.q[0]